src/OSPSuite.Core/Domain/Services/ResidualCalculator.cs

```diff
@@ -34,13 +34,28 @@ public ResidualsResult Calculate(IReadOnlyList<SimulationRunResults> simulations
       {
          if (simulationsResults.Any(x => !x.Success))
             return residualResultWithErrorFrom(simulationsResults);
-        
+
          var simulationColumnsCache = new Cache<string, DataColumn>(x => x.PathAsString, x => null);
          simulationsResults.Select(x => x.Results).Each(repo => simulationColumnsCache.AddRange(repo.AllButBaseGrid()));
 
-         var residualResult = new ResidualsResult();
+         return calculateResidualsResult(allOutputMappings, simulationColumnsCache);
+      }
 
-         foreach (var outputMapping in allOutputMappings)
+      public ResidualsResult Calculate(DataRepository simulationResultRepository, IReadOnlyList<OutputMapping> allOutputMappings)
+      {
+         var simulationColumnsCache = new Cache<string, DataColumn>(x => x.PathAsString, x => null);
+         simulationColumnsCache.AddRange(simulationResultRepository.AllButBaseGrid());
+
+         return calculateResidualsResult(allOutputMappings, simulationColumnsCache);
+      }
+      private ResidualsResult calculateResidualsResult(IReadOnlyList<OutputMapping> allOutputMappings,
```

Comment:
@georgeDaskalakis Please run resharper in this file. It looks like its completely skewed